rfc/0000-configuration-and-shared-files.md

```diff
@@ -0,0 +1,296 @@
+  - `data_home`: The directory where shared files are stored. Equivalent to `PACT_DATA_HOME`.
+  - `do_not_track`: If set to `true`, the Pact implementation _must not_ report any usage analytics data back to the Pact Foundation.
+
+- `broker`
+
+  To store configuration as to how to connect to the Pact Broker:
+
+  ```toml
+  [broker]
+  base_url = "https://pact-broker.local.company.com:9292"
+  username = "user"
+  password = "password"
+  ```
+
+  The following entries are proposed:
+
+  - `base_url`: The base URL of the Pact Broker. Equivalent to `PACT_BROKER_BASE_URL`.
+  - `username`: The username to use when authenticating with the Pact Broker. Equivalent to `PACT_BROKER_USERNAME`.
+  - `password`: The password to use when authenticating with the Pact Broker. Equivalent to `PACT_BROKER_PASSWORD`.
+  - `token`: The token to use when authenticating with the Pact Broker. Equivalent to `PACT_BROKER_TOKEN`.
+
+  The Pact Broker itself _may_ have additional configuration options within the `broker` table.
+
+- `{ident}`
+
+  To store configuration specific to a particular language implementation or tool. It is up to the implementation or tool to define the meaning of these entries, and each implementation or tool should only read entries under its own identifier.
+
+  Sub-tables may used to group configuration data within each `{ident}` section.
+
+  ```toml
+  [ruby]
+  bar = "baz"
+
+  [net.foo]
+  disabled = true
+
+  [pactflow]
+  enable_other_feature = false
+  ```
+
+#### Path Handling
+
+To ensure a uniform experience across the Pact ecosystem and operating systems, the following rules _must_ be followed when handling paths in the configuration file:
+
+- Paths _must_ be stored using the forward slash (`/`) as the path separator.
+  - This avoids the need to escape backslashes (`\`) in Windows paths.
+  - If a backslash is used, it is undefined behaviour. The implementation _may_ attempt to convert the backslashes to forward slashes, or may use the value as-is.
+- The tilde (`~`) character _must_ be expanded to the user's home directory if it is the first character in a path.
+  - For example, parsing `~/.local/share/pact` should result in `/home/user/.local/share/pact` on Unix-like systems.
+  - If the tilde appears elsewhere in the path, it is treated as a literal character.
+  - If someone wishes to use a literal tilde for a relative path, they _must_ use `./~` instead.
+- Relative paths are allowed, and _must_ be resolved relative to the configuration file.
+  - For example, if the configuration file is located at `/workdir/project/.pact.toml`, and a value is set to `.pact/data`, the resolved path should be `/workdir/project/.pact/data`.
+
+
+### Shared Files
+
+At present, there are only a few files which are shared across the Pact ecosystem. These are:
+
+- Plugins, currently stored in `~/.pact/plugins`
+- The Pact core library (i.e., the Pact FFI), though only in some implementations.
```

Comment:
what are proposed paths? e.g. `~/.local/share/pact/ffi` and fallback `~/.pact/ffi` ?

Reply:
I don't actually specify how the FFI binaries/headers should be stored, mostly because I don't know what practices are used across the ecosystem.

Having `$PACT_DATA_HOME/ffi/pact.h` and `$PACT_DATA_HOME/ffi/(lib)pact.(ext)` seems quite reasonable to me though.

This assumes it is dynamically linked though, which isn't the case for all implementations.

Reply:
We might need to think about versioning here. For instance, a user might be using Pact JS and Pact Go. It might be OK to have a latter version of the FFI for one and not the other (e.g. they have just started using Pact Go, and are on the latest version, but Pact JS has an existing suite and they are on an older version). In theory, this should only be a problem for major versions of the FFI, but in practice this isn't always the case.

Reply:
Or perhaps a way to address it is by convention. 

e.g. 

* [base path]/ffi/pact.h
* [base path]/ffi/pact.so
* [base path]/ffi/pact-js/pact.h
* [base path]/ffi/pact-js/pact.so
* [base path]/ffi/pact-go/pact.h
* [base path]/ffi/pact-go/pact.so

Pact JS and Pact Go would have separate versions, and all other libraries use the fallback

Reply:
I was actually thinking more about this, and I was wondering whether we should recommend the standard Unix way of organising these files:

```
$PACT_DATA_HOME/include/pact.h
$PACT_DATA_HOME/lib/libpact.so
```

and then libraries can opt in to library-specific version of the FFI with

```
$PACT_DATA_HOME/include/pact.js.h
$PACT_DATA_HOME/lib/libpact.js.so
$PACT_DATA_HOME/include/pact.go.h
$PACT_DATA_HOME/lib/libpact.go.so
...
```

This follows the same pattern of having multiple version of libraries presented as:

```
lib/libfoo.so      -> libfoo.2.so
lib/libfoo.2.so    -> libfoo.2.1.so
lib/libfoo.2.1.so
lib/libfoo.1.so    -> libfoo.1.8.so
lib/libfoo.1.8.so
```

(the arrows are symlinks, but we need not use symlinks here)